ledger/apply/asset.go

```diff
@@ -308,39 +337,51 @@ func AssetTransfer(ct transactions.AssetTransferTxnFields, header transactions.H
 			return fmt.Errorf("cannot close asset by clawback")
 		}
 
-		// Fetch the sender balance record. We will use this to ensure
-		// that the sender is not the creator of the asset, and to
-		// figure out how much of the asset to move.
-		snd, err := balances.Get(source, false)
+		record, err := balances.Get(source, false)
 		if err != nil {
 			return err
 		}
 
+		if record.TotalAssets == 0 {
+			return fmt.Errorf("cannot close asset holding on account %s : account is not opted in asset %d ", source.String(), ct.XferAsset)
+		}
+
+		// Fetch the sender asset data. We will use this to ensure
+		// that the sender is not the creator of the asset, and to
+		// figure out how much of the asset to move.
+
 		// The creator of the asset cannot close their holding of the
 		// asset. Check if we are the creator by seeing if there is an
 		// AssetParams entry for the asset index.
-		if _, ok := snd.AssetParams[ct.XferAsset]; ok {
+		ok, err := balances.HasAssetParams(source, ct.XferAsset)
+		if err != nil {
+			return err
+		}
+		if ok {
 			return fmt.Errorf("cannot close asset ID in allocating account")
 		}
 
 		// Fetch our asset holding, which should exist since we're
 		// closing it out
-		sndHolding, ok := snd.Assets[ct.XferAsset]
+		sndHolding, ok, err := balances.GetAssetHolding(source, ct.XferAsset)
+		if err != nil {
+			return err
+		}
 		if !ok {
 			return fmt.Errorf("asset %v not present in account %v", ct.XferAsset, source)
 		}
 
-		// Fetch the destination balance record to check if we are
+		// Fetch the destination asset params to check if we are
 		// closing out to the creator
-		dst, err := balances.Get(ct.AssetCloseTo, false)
+		dstAssetParamsExist, err := balances.HasAssetParams(ct.AssetCloseTo, ct.XferAsset)
```

Comment:
This makes me think balances.HasAssetParams() should have an optimization for asking about the zero address.  That will be the most common case by far, and the answer is always False.

Reply:
since this is a non-issue, I'm going to resolve this one. Please let me know if I missed anything.